File: oo/src/test/java/oo/ep/M2Test.java
package oo.ep;

import org.junit.Assert;
import org.junit.Test;

public class M2Test {

    public interface TestTemplate extends M1Test.TestTemplate {
        default void test() {
            M1Test.TestTemplate.super.test();

            Exp expr1 = new Add(new Lit(1.0), new Lit(2.0));
            Assert.assertEquals("(1.0+2.0)", expr1.prettyp());

            Exp expr2 = new Lit(2.0);
            Assert.assertEquals("2.0", expr2.prettyp());

            Assert.assertEquals("(1.0-2.0)", new Sub(new Lit(1.0), new Lit(2.0)).prettyp());
            Assert.assertEquals("((1.0-2.0)+(5.0+6.0))", new Add(new Sub(new Lit(1.0), new Lit(2.0)), new Add(new Lit(5.0), new Lit(6.0))).prettyp());
        }
    }
    private static class ActualTest implements TestTemplate {}

    @Test
    public void testTest() { new ActualTest().test(); }
}

File: coco_scala/src/test/scala/exp/m2/M2Test.scala
  package exp.m2

  import org.scalatest.FunSuite

  trait TestTemplate[FT] extends Factory[FT] with exp.m1.TestTemplate[FT] {
    val suite:FunSuite
    import suite._

    override def test() : Unit = {
      super.test()

     val expr1 = this.add(this.lit(1.0), this.lit(2.0))
     assert("(1.0+2.0)" === expr1.prettyp)

     val expr2 = this.lit(2.0)
     assert("2.0" === expr2.prettyp)

     assert("(1.0-2.0)" === convert(this.sub(this.lit(1.0), this.lit(2.0))).prettyp)
     assert( "((1.0-2.0)+(5.0+6.0))" === convert(this.add(this.sub(this.lit(1.0), this.lit(2.0)), this.add(this.lit(5.0), this.lit(6.0)))).prettyp)
    }
  }

  class M2Test extends FunSuite { self =>

    object ActualTest extends TestTemplate[exp.m2.finalized.Exp] with finalized.Factory {
      val suite: FunSuite = self
    }

    test("M2") {
      ActualTest.test()
    }
  }
